File: Data Structure/src/python/sorting/bubble_sort.py
def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]


def bubble_sort(arr):
    swap_counter = 0
    for i in range(len(arr) - 1, 1, -1):
        swapped = False
        for j in range(i):
            if arr[j] > arr[j + 1]:
                swap(arr, j, j + 1)
                swapped = True
                swap_counter += 1

        if not swapped:
            break

    print(f"Total Swaps: {swap_counter}")


if __name__ == "__main__":
    arr = [64, 34, 25, 12, 22, 11, 90]
    print("Unsorted array:", arr)
    bubble_sort(arr)
    print("Sorted array:", arr)

    arr = [i for i in range(10)]
    print("Unsorted array:", arr)
    bubble_sort(arr)
    print("Sorted array:", arr)

File: Data Structure/src/go/sorting/bubble_sort.go
package main

import "fmt"

func bubble_sort(arr []int, n int) {
	swap_counter := 0
	swapped := false
	for i := n - 1; i > 1; i-- {
		for j := 0; j < i; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
				swap_counter++
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
	fmt.Println("Number of swaps : ", swap_counter)

}

func main() {
	arr := []int{64, 34, 25, 12, 22, 11, 90}
	n := len(arr)
	fmt.Println("Unsorted array:", arr)
	bubble_sort(arr, n)
	fmt.Println("Sorted array:", arr)

	arr2 := []int{1, 2, 3, 3, 4, 5, 6}
	n2 := len(arr)
	fmt.Println("Unsorted array:", arr)
	bubble_sort(arr2, n2)
	fmt.Println("Sorted array:", arr)
}
